File: Examples/FurEdVB/MainForm.vb
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports Furcadia.Drawing.Graphics


Partial Public Class MainForm
    Inherits Form
    Public Sub New()
        InitializeComponent()
    End Sub

    Private lastINIPath As String = ""

  

    Private Sub loadButton_Click(sender As Object, e As EventArgs) Handles loadButton.Click
        LoadINI()
    End Sub

    Private Function LoadINI() As Boolean
        If chooseIniFileDialog.ShowDialog() <> DialogResult.OK Then
            Return False
        End If

        Dim IniText As String() = System.IO.File.ReadAllLines(chooseIniFileDialog.FileName)
        For Each line As String In IniText
            If line.StartsWith("Colors=") Then
                LoadColourCode(line.Substring(7))
            End If
            If line.StartsWith("Desc=") Then
                TxtBxDescription.Text = line.Substring(5)
            End If

        Next

        lastINIPath = chooseIniFileDialog.FileName
        saveButton.Enabled = True

        Return True
    End Function

    Private Sub saveButton_Click(sender As Object, e As EventArgs) Handles saveButton.Click
        saveIniFileDialog.InitialDirectory = System.IO.Path.GetDirectoryName(lastINIPath)
        saveIniFileDialog.FileName = System.IO.Path.GetFileName(lastINIPath)

        If saveIniFileDialog.ShowDialog() = DialogResult.OK Then
            Dim IniText As String() = System.IO.File.ReadAllLines(chooseIniFileDialog.FileName)
            Dim idx As Integer = 0
            Dim linenum As Integer = -1
            For Each line As String In IniText
                If line.StartsWith("Colors=") Then
                    linenum = idx
                    Exit For
                End If
                idx += 1
            Next
            If linenum <> -1 Then
                IniText(linenum) = "Colors=" & ColourCode & SpeciesBlock
            End If
            System.IO.File.WriteAllLines(saveIniFileDialog.FileName, IniText)
        End If
    End Sub

    Private ColourCode As String
    Private SpeciesBlock As String

    Private Function LoadColourCode(colourcode__1 As String) As Boolean
        If colourcode__1.Length <> 14 OrElse colourcode__1(0) <> "t"c Then
            System.Windows.Forms.MessageBox.Show("Invalid colour code!")
            Return False
        End If

        ColourCode = colourcode__1.Substring(0, 11)
        SpeciesBlock = colourcode__1.Substring(11)
        'FurrePreview1.LoadStart(AscW(colourcode__1.Chars(12)) - 35, AscW(colourcode__1.Chars(11)) - 35)
        CustomFurrePreview1.LoadStart(AscW(colourcode__1.Chars(12)) - 35, AscW(colourcode__1.Chars(11)) - 35)
        Dim FurColour As Integer = AscW(colourcode__1.Chars(1)) - 35
        Dim MarkingsColour As Integer = AscW(colourcode__1.Chars(2)) - 35
        Dim HairColour As Integer = AscW(colourcode__1.Chars(3)) - 35
        Dim EyeColour As Integer = AscW(colourcode__1.Chars(4)) - 35
        Dim BadgeColour As Integer = AscW(colourcode__1.Chars(5)) - 35
        Dim VestColour As Integer = AscW(colourcode__1.Chars(6)) - 35
        Dim BracerColour As Integer = AscW(colourcode__1.Chars(7)) - 35
        Dim CapeColour As Integer = AscW(colourcode__1.Chars(8)) - 35
        Dim BootColour As Integer = AscW(colourcode__1.Chars(9)) - 35
        Dim TrousersColour As Integer = AscW(colourcode__1.Chars(10)) - 35

        UpdatingColours = True
        furListBox.SelectedIndex = FurColour
        markingListBox.SelectedIndex = MarkingsColour
        hairListBox.SelectedIndex = HairColour
        eyeListBox.SelectedIndex = EyeColour
        badgeListBox.SelectedIndex = BadgeColour
        vestListBox.SelectedIndex = VestColour
        bracerListBox.SelectedIndex = BracerColour
        capeListBox.SelectedIndex = CapeColour
        bootsListBox.SelectedIndex = BootColour
        trouserListBox.SelectedIndex = TrousersColour
        UpdatingColours = False

        'FurrePreview1.UpdateColourCode(ColourCode)
        CustomFurrePreview1.UpdateColourCode(ColourCode)
        colourCodeTextBox.Text = ColourCode & SpeciesBlock
        Return True
    End Function

    Private UpdatingColours As Boolean = False

    Private Sub RenderColourChoice(sender As Object, e As DrawItemEventArgs) _
        Handles badgeListBox.DrawItem, bootsListBox.DrawItem, bracerListBox.DrawItem, _
        capeListBox.DrawItem, eyeListBox.DrawItem, furListBox.DrawItem, hairListBox.DrawItem, _
        markingListBox.DrawItem, trouserListBox.DrawItem, vestListBox.DrawItem
        'bool big = !(sender == badgeListBox || sender == hairListBox);
        Dim big As Boolean = True

        e.DrawBackground()
        e.DrawFocusRectangle()

        Dim giver As ListBox = TryCast(sender, ListBox)
        Dim TextArea As Rectangle = e.Bounds
        TextArea.X += (If(big, 68, 34))
        'e.Graphics.DrawString((string)giver.Items[e.Index], e.Font, new SolidBrush(e.ForeColor), e.Bounds.Location);
        TextRenderer.DrawText(e.Graphics, DirectCast(giver.Items(e.Index), String), e.Font, TextArea, e.ForeColor, e.BackColor, _
         TextFormatFlags.Left)

        Dim Target As Byte() = Nothing

        If giver.Name = badgeListBox.Name Then
            Target = New Byte(7) {}
            For i As Integer = 0 To 7
                Target(i) = Remapper.BadgeRemap(e.Index)
            Next
        ElseIf giver.Name = bracerListBox.Name Then
            Target = Remapper.BracersRemap(e.Index)
        ElseIf giver.Name = bootsListBox.Name Then
            Target = Remapper.BootRemap(e.Index)
        ElseIf giver.Name = capeListBox.Name Then
            Target = Remapper.CapeRemap(e.Index)
        ElseIf giver.Name = eyeListBox.Name Then
            Target = New Byte(7) {}
            For i As Integer = 0 To 7
                Target(i) = Remapper.EyeRemap(e.Index)
            Next
        ElseIf giver.Name = furListBox.Name Then
            Target = Remapper.FurRemap(e.Index)
        ElseIf giver.Name = hairListBox.Name Then
            Target = Remapper.HairRemap(e.Index)
        ElseIf giver.Name = markingListBox.Name Then
            Target = Remapper.MarkingsRemap(e.Index)
        ElseIf giver.Name = trouserListBox.Name Then
            Target = Remapper.TrousersRemap(e.Index)
        ElseIf giver.Name = vestListBox.Name Then
            Target = Remapper.VestRemap(e.Index)
        End If

        Dim dX As Integer = e.Bounds.X + 2
        Dim dY As Integer = e.Bounds.Y + 3
        'For i As Integer = 0 To 7
        '    e.Graphics.FillRectangle(New SolidBrush(FurrePreview.Palette(Target(i))), dX, dY, (If(big, 8, 4)), 8)
        '    dX += (If(big, 8, 4))
        'Next
        For i As Integer = 0 To 7
            e.Graphics.FillRectangle(New SolidBrush(CustomFurrePreview1.Palette(Target(i))), dX, dY, (If(big, 8, 4)), 8)
            dX += (If(big, 8, 4))
        Next
        Dim BorderBrush As Brush = New SolidBrush(e.BackColor)
        e.Graphics.FillRectangle(BorderBrush, e.Bounds.X + 2, e.Bounds.Y + 3, 1, 1)
        e.Graphics.FillRectangle(BorderBrush, e.Bounds.X + 2, e.Bounds.Y + 10, 1, 1)
        e.Graphics.FillRectangle(BorderBrush, e.Bounds.X + (If(big, 65, 33)), e.Bounds.Y + 3, 1, 1)
        e.Graphics.FillRectangle(BorderBrush, e.Bounds.X + (If(big, 65, 33)), e.Bounds.Y + 10, 1, 1)
    End Sub

    Private Function GetNumberForListBox(giver As ListBox) As Integer
        If giver.Name = furListBox.Name Then
            Return 0
        ElseIf giver.Name = markingListBox.Name Then
            Return 1
        ElseIf giver.Name = hairListBox.Name Then
            Return 2
        ElseIf giver.Name = eyeListBox.Name Then
            Return 3
        ElseIf giver.Name = badgeListBox.Name Then
            Return 4
        ElseIf giver.Name = vestListBox.Name Then
            Return 5
        ElseIf giver.Name = bracerListBox.Name Then
            Return 6
        ElseIf giver.Name = capeListBox.Name Then
            Return 7
        ElseIf giver.Name = bootsListBox.Name Then
            Return 8
        ElseIf giver.Name = trouserListBox.Name Then
            Return 9
        End If

        Return -1
    End Function

    Private Sub ChangeColourChoice(sender As Object, e As EventArgs) _
        Handles badgeListBox.SelectedIndexChanged, bootsListBox.SelectedIndexChanged, bracerListBox.SelectedIndexChanged, _
        capeListBox.SelectedIndexChanged, eyeListBox.SelectedIndexChanged, furListBox.SelectedIndexChanged, hairListBox.SelectedIndexChanged, _
        markingListBox.SelectedIndexChanged, trouserListBox.SelectedIndexChanged, vestListBox.SelectedIndexChanged

        If UpdatingColours Then
            Return
        End If

        Dim giver As ListBox = TryCast(sender, ListBox)
        Dim modify As Integer = GetNumberForListBox(giver) + 1

        ColourCode = ColourCode.Substring(0, modify) & (ChrW(giver.SelectedIndex + 35)).ToString() & ColourCode.Substring(modify + 1)
        'FurrePreview1.UpdateColourCode(ColourCode)
        CustomFurrePreview1.UpdateColourCode(ColourCode)
        colourCodeTextBox.Text = ColourCode & SpeciesBlock
    End Sub

    Private Sub button1_Click(sender As Object, e As EventArgs) Handles loadButton.Click
        LoadColourCode(colourCodeTextBox.Text)
    End Sub

    Private Sub aboutButton_Click(sender As Object, e As EventArgs) Handles aboutButton.Click

        MessageBox.Show("Treeki's Extended Colour Code Editor" & vbLf & vbLf & "Originally created October 2009." & vbLf & "Updated and re-released December 2011" & vbLf & vbLf & "Thanks to Kylix for the FOX code which I cannibalised :p", "About")
    End Sub

    ' Badge
    ' Bracers
    ' Boots
    ' Cape
    ' Eyes
    ' Fur
    ' Hair
    ' Markings
    ' Trousers
    ' Vest
    Shared tabIndicesToNumbers As Integer() = New Integer() {
        4,
        6,
        8,
        7,
        3,
        0,
        2,
        1,
        9,
        5}

    Private Sub tabControl1_SelectedIndexChanged1(sender As Object, e As System.EventArgs) Handles tabControl1.SelectedIndexChanged
        Dim number As Integer = tabIndicesToNumbers(tabControl1.SelectedIndex)

        'FurrePreview1.CurrentRemap = number
        CustomFurrePreview1.CurrentRemap = number
    End Sub

    Private Sub MainForm_Load1(sender As Object, e As System.EventArgs) Handles Me.Load
        Furcpath = New Furcadia.IO.Paths()

        'FurrePreview.Prepare()
        CustomFurrePreview1.Prepare()

        chooseIniFileDialog.InitialDirectory = FurcPath.GetFurcadiaDocPath + "Furcadia Characters"



        LoadColourCode("t#############")
        saveButton.Enabled = False


        'LoadColourCode("t89J,I@,,&;$%#");

        tabControl1.SelectTab(5)
    End Sub

    Private Sub DreamToolStripMenuItem_Click(sender As System.Object, e As System.EventArgs) Handles DreamToolStripMenuItem.Click

    End Sub

    Private Sub Button1_Click_1(sender As System.Object, e As System.EventArgs) Handles Button1.Click
        With PatchBrowser
            .SelectedPath = FurcPath.GetFurcadiaDocPath()
            If .ShowDialog = Windows.Forms.DialogResult.OK Then
                CustomFurrePreview1.PatchPath = .SelectedPath
                CustomFurrePreview1.Prepare()
            End If
        End With
    End Sub
End Class

File: Examples/CCEditor/FurEd/MainForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;

namespace FurEd {
    public partial class MainForm : Form {
        public MainForm() {
            InitializeComponent();
        }

		string lastINIPath = "";

        private void MainForm_Load(object sender, EventArgs e) {
            Remapper.Prepare();

			string path = String.Join(
				System.IO.Path.DirectorySeparatorChar.ToString(),
				new string[] {
					Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments),
					"Furcadia", "Furcadia Characters"
				});

			if (System.IO.Directory.Exists(path))
				chooseIniFileDialog.InitialDirectory = path;

            if (!LoadINI()) {
                LoadColourCode("t#############");
                saveButton.Enabled = false;
            }

            //LoadColourCode("t89J,I@,,&;$%#");

			tabControl1.SelectTab(5);
        }

		private void loadButton_Click(object sender, EventArgs e) {
			LoadINI();
		}

		private bool LoadINI() {
			if (chooseIniFileDialog.ShowDialog() != DialogResult.OK)
				return false;

			string[] IniText = System.IO.File.ReadAllLines(chooseIniFileDialog.FileName);
			foreach (string line in IniText) {
				if (line.StartsWith("Colors=")) {
					LoadColourCode(line.Substring(7));
				}
			}

			lastINIPath = chooseIniFileDialog.FileName;
			saveButton.Enabled = true;

			return true;
		}

        private void button4_Click(object sender, EventArgs e) {
			saveIniFileDialog.InitialDirectory = System.IO.Path.GetDirectoryName(lastINIPath);
			saveIniFileDialog.FileName = System.IO.Path.GetFileName(lastINIPath);

            if (saveIniFileDialog.ShowDialog() == DialogResult.OK) {
                string[] IniText = System.IO.File.ReadAllLines(chooseIniFileDialog.FileName);
                int idx = 0;
                int linenum = -1;
                foreach (string line in IniText) {
                    if (line.StartsWith("Colors=")) {
                        linenum = idx;
                        break;
                    }
                    idx++;
                }
                if (linenum != -1) {
                    IniText[linenum] = "Colors=" + ColourCode + SpeciesBlock;
                }
                System.IO.File.WriteAllLines(saveIniFileDialog.FileName, IniText);
            }
        }

        private string ColourCode;
        private string SpeciesBlock;

        private bool LoadColourCode(string colourcode) {
            if (colourcode.Length != 14 || colourcode[0] != 't') {
                System.Windows.Forms.MessageBox.Show("Invalid colour code!");
                return false;
            }

            ColourCode = colourcode.Substring(0, 11);
            SpeciesBlock = colourcode.Substring(11);

            furrePreview1.LoadStart(colourcode[12] - 35, colourcode[11] - 35);
            furrePreview2.LoadStart(colourcode[12] - 35, colourcode[11] - 35);
            furrePreview3.LoadStart(colourcode[12] - 35, colourcode[11] - 35);

            int FurColour = colourcode[1] - 35;
            int MarkingsColour = colourcode[2] - 35;
            int HairColour = colourcode[3] - 35;
            int EyeColour = colourcode[4] - 35;
            int BadgeColour = colourcode[5] - 35;
            int VestColour = colourcode[6] - 35;
            int BracerColour = colourcode[7] - 35;
            int CapeColour = colourcode[8] - 35;
            int BootColour = colourcode[9] - 35;
            int TrousersColour = colourcode[10] - 35;

            UpdatingColours = true;
            furListBox.SelectedIndex = FurColour;
            markingListBox.SelectedIndex = MarkingsColour;
            hairListBox.SelectedIndex = HairColour;
            eyeListBox.SelectedIndex = EyeColour;
            badgeListBox.SelectedIndex = BadgeColour;
            vestListBox.SelectedIndex = VestColour;
            bracerListBox.SelectedIndex = BracerColour;
            capeListBox.SelectedIndex = CapeColour;
            bootsListBox.SelectedIndex = BootColour;
            trouserListBox.SelectedIndex = TrousersColour;
            UpdatingColours = false;

            furrePreview1.UpdateColourCode(ColourCode);
            furrePreview2.UpdateColourCode(ColourCode);
            furrePreview3.UpdateColourCode(ColourCode);

            colourCodeTextBox.Text = ColourCode + SpeciesBlock;
            return true;
        }

        private bool UpdatingColours = false;

        private void RenderColourChoice(object sender, DrawItemEventArgs e) {
			//bool big = !(sender == badgeListBox || sender == hairListBox);
			bool big = true;

            e.DrawBackground();
            e.DrawFocusRectangle();

            ListBox giver = sender as ListBox;
            Rectangle TextArea = e.Bounds;
            TextArea.X += (big ? 68 : 34);
            //e.Graphics.DrawString((string)giver.Items[e.Index], e.Font, new SolidBrush(e.ForeColor), e.Bounds.Location);
            TextRenderer.DrawText(e.Graphics, (string)giver.Items[e.Index], e.Font, TextArea, e.ForeColor, e.BackColor, TextFormatFlags.Left);

            byte[] Target = null;

            if (giver == badgeListBox) {
                Target = new byte[8];
                for (int i = 0; i < 8; i++) {
                    Target[i] = Remapper.BadgeRemap[e.Index];
                }
            } else if (giver == bracerListBox) {
                Target = Remapper.BracersRemap[e.Index];
            } else if (giver == bootsListBox) {
                Target = Remapper.BootRemap[e.Index];
            } else if (giver == capeListBox) {
                Target = Remapper.CapeRemap[e.Index];
            } else if (giver == eyeListBox) {
                Target = new byte[8];
                for (int i = 0; i < 8; i++) {
                    Target[i] = Remapper.EyeRemap[e.Index];
                }
            } else if (giver == furListBox) {
                Target = Remapper.FurRemap[e.Index];
            } else if (giver == hairListBox) {
                Target = Remapper.HairRemap[e.Index];
            } else if (giver == markingListBox) {
                Target = Remapper.MarkingsRemap[e.Index];
            } else if (giver == trouserListBox) {
                Target = Remapper.TrousersRemap[e.Index];
            } else if (giver == vestListBox) {
                Target = Remapper.VestRemap[e.Index];
            }

            int dX = e.Bounds.X + 2;
            int dY = e.Bounds.Y + 3;
            for (int i = 0; i < 8; i++) {
				e.Graphics.FillRectangle(new SolidBrush(Remapper.Palette[Target[i]]), dX, dY, (big ? 8 : 4), 8);
                dX += (big ? 8 : 4);
            }

            Brush BorderBrush = new SolidBrush(e.BackColor);
            e.Graphics.FillRectangle(BorderBrush, e.Bounds.X + 2, e.Bounds.Y + 3, 1, 1);
            e.Graphics.FillRectangle(BorderBrush, e.Bounds.X + 2, e.Bounds.Y + 10, 1, 1);
            e.Graphics.FillRectangle(BorderBrush, e.Bounds.X + (big ? 65 : 33), e.Bounds.Y + 3, 1, 1);
			e.Graphics.FillRectangle(BorderBrush, e.Bounds.X + (big ? 65 : 33), e.Bounds.Y + 10, 1, 1);
        }

		private int GetNumberForListBox(ListBox giver) {
			if (giver == furListBox) {
				return 0;
			} else if (giver == markingListBox) {
				return 1;
			} else if (giver == hairListBox) {
				return 2;
			} else if (giver == eyeListBox) {
				return 3;
			} else if (giver == badgeListBox) {
				return 4;
			} else if (giver == vestListBox) {
				return 5;
			} else if (giver == bracerListBox) {
				return 6;
			} else if (giver == capeListBox) {
				return 7;
			} else if (giver == bootsListBox) {
				return 8;
			} else if (giver == trouserListBox) {
				return 9;
			}

			return -1;
		}

        private void ChangeColourChoice(object sender, EventArgs e) {
            if (UpdatingColours) return;

            ListBox giver = sender as ListBox;
            int modify = GetNumberForListBox(giver) + 1;

            ColourCode = ColourCode.Substring(0,modify) + ((char)(giver.SelectedIndex + 35)).ToString() + ColourCode.Substring(modify+1);
            furrePreview1.UpdateColourCode(ColourCode);
            furrePreview2.UpdateColourCode(ColourCode);
            furrePreview3.UpdateColourCode(ColourCode);

            colourCodeTextBox.Text = ColourCode + SpeciesBlock;
        }

        private void button1_Click(object sender, EventArgs e) {
            LoadColourCode(colourCodeTextBox.Text);
        }

		private void aboutButton_Click(object sender, EventArgs e) {
			MessageBox.Show(
				"Treeki's Extended Colour Code Editor\n\n" +
				"Originally created October 2009.\n" +
				"Updated and re-released December 2011\n\n" +
				"Thanks to Kylix for the FOX code which I cannibalised :p",

				"About");
		}

		static int[] tabIndicesToNumbers = new int[] {
			4, // Badge
			6, // Bracers
			8, // Boots
			7, // Cape
			3, // Eyes
			0, // Fur
			2, // Hair
			1, // Markings
			9, // Trousers
			5, // Vest
		};

		private void tabControl1_SelectedIndexChanged(object sender, EventArgs e) {
			int number = tabIndicesToNumbers[tabControl1.SelectedIndex];

			furrePreview1.CurrentRemap = number;
			furrePreview2.CurrentRemap = number;
			furrePreview3.CurrentRemap = number;
		}
    }
}
